jwst/tests/base_test.py

```diff
@@ -0,0 +1,153 @@
+                line = str(item)
+                if 'href' in line and '..' not in line:
+                    start = line.index('"')+1
+                    end = line.rindex('"')
+                    fname = line[start:end]
+                    
+                    self.get_data(*args,fname, docopy=True)
+        else:
+            # Make copy from local cache
+            cfg_names = glob.glob(os.path.join(cfg_dir,'*'))
+            for name in cfg_names:
+                filename = os.path.basename(name)
+                self.get_data(*args,filename, docopy=True)
+
+        # Move back to original directory
+        os.chdir(test_dir)
+
+
+# Pytest function to support the parameterization of these classes
+def pytest_generate_tests(metafunc):
+    # called once per each test function
+    funcarglist = metafunc.cls.params[metafunc.function.__name__]
+    argnames = sorted(funcarglist[0])
+    idlist = [funcargs['id'] for funcargs in funcarglist]
+    del argnames[argnames.index('id')]
+    metafunc.parametrize(argnames, [[funcargs[name] for name in argnames]
+            for funcargs in funcarglist], ids=idlist)
+
+
+@pytest.mark.usefixtures('_jail')
+class BaseJWSTTestSteps(BaseJWSTTest):
+
+    params = {'test_steps':[dict(input="",
+                                 test_dir=None,
+                                 step_class=None,
+                                 step_pars=dict(),
+                                 output_truth="",
+                                 output_hdus=[])
+                            ]
+             }
+
+    def test_steps(self, input, test_dir, step_class, step_pars,
+                   output_truth, output_hdus):
+        """
+        Template method for parameterizing all the tests of MIRI pipeline
+        processing steps.
+        """
+        if test_dir is None:
+            return
+
+        self.test_dir = test_dir
+        self.ref_loc = [self.test_dir, 'truth']
+
+        # can be removed once all truth files have been updated
+        self.ignore_keywords += ['FILENAME']
+
+        input_file = self.get_data(self.test_dir, input)
+
+        result = step_class.call(input_file, **step_pars)
+
+        output_file = result.meta.filename
+        result.save(output_file)
+        result.close()
+
+        output_pars = None
+        if isinstance(output_truth, tuple):
+            output_pars = output_truth[1]
+            output_truth = output_truth[0]
+
+        if not output_pars:
+            if output_hdus:
+                output_spec = (output_file, output_truth, output_hdus)
+            else:
+                output_spec = (output_file, output_truth)
+        else:
+            output_spec = {'files':(output_file, output_truth),
+                           'pars':output_pars}
+        outputs = [output_spec]
+        self.compare_outputs(outputs)
+
+
+# function(s) for tests written using functional form, not as BaseTest class tests
+def raw_from_asn(asn_file):
```

Comment:
There is already such functionality in `ci-watson` for HST. Are you able to re-use that for JWST?

Reply:
JWST association files are different I believe.

Reply:
Perhaps a future PR but I think it might benefit more people if JWST's version of `raw_from_asn` is moved to `ci-watson`. Such functionality is useful beyond `jwst` pipeline testing. Actually, maybe for both HST and JWST, they are useful beyond `ci-watson` but I am hesitant to suggest `stsci.tools` without giving it more thought.